docs/usage/jssengine.md

```diff
@@ -0,0 +1,521 @@
+ - Allow forced expiration of a session as long as the `SSLEngine`'s
+   connection isn't yet closed.
+ - Report accurate creation/expiration/last accessed times.
+
+However, other features of sessions (such as configuring location and size of
+the session cache) aren't yet configurable.
+
+
+## Design of the `JSSEngine`
+
+### Class Structuring
+
+The below is a digram showing the structure of `JSSEngine` classes:
+
+                           -----------
+                          | JSSEngine |-------------------------
+                           -----------                          \
+                            /       \                            \
+     ------------------------       ------------------------     ------
+    | JSSEngineReferenceImpl |     | JSSEngineOptimizedImpl |   | .... |
+     ------------------------       ------------------------     ------
+
+`JSSEngine` is an abstract class extending [`SSLEngine`][javax.ssl-engine].
+This class implements some of the boilerplate required for implementing a
+`SSLEngine`, including handling cipher and protocol version configuration.
+Individual implementations implement `wrap`, `unwrap`, and whatever specifics
+are necessary to initialize and release the SSL-backed `PRFileDesc`.
+
+We expect two primary implementations:
+
+ - `JSSEngineReferenceImpl`, a reference implementation with more logging
+   and debugging statements. This also includes port-based debugging, so
+   situations where a `SSLEngine` is without writing to the network can
+   still be tracked and analyzed in Wireshark. Each call to `wrap` or `unwrap`
+   makes several JNI calls, incurring lots of overhead.
+ - `JSSEngineOptimizedImpl`, an optimized, production-ready implementation.
+   This one is harder to debug due to fewer logging statements, but does
+   improve performance significantly with fewer JNI calls.
+
+
+### Non-Blocking IO
+
+NSPR introduces a platform-independent abstraction over C's file descriptors
+(usually an `int`) in the form of the `PRFileDesc` structure. This is
+layer-able, allowing us to write our own and then have it be accepted by
+NSS's SSL `PRFileDesc` implementation as the underlying transport.
+
+Our `PRFileDesc` is called `BufferPRFD` and lives next to the `JSSEngine`
+implementation in `org/mozilla/jss/ssl/javax`. Each `BufferPRFD` is backed
+by two `j_buffer` instances (located adjacent). A `j_buffer` is a circular
+ring buffer optimized to allow efficient access to the underlying data. One
+`j_buffer` is dedicated to the read end of this `BufferPRFD` (the data that
+is returned when `recv(...)` is called); the other is dedicated to the write
+end of this `BufferPRFD` (the data that is waiting to get sent on the wire).
+Note that the `j_buffer` instances exist independently from the `BufferPRFD`:
+the `JSSEngine` itself creates the `j_buffer`s and hands them to the
+`BufferPRFD` to use, because it too needs to be able to read from them.
+
```

Comment:
I found it interesting when looking at the code that the NSPR sockets were being used to wrap or  unwrap locally and then handing the data off to wrap or unwrap. Would a simple statement to that at a high level make the whole thing a bit clearer to the reader? I know that all that gets explained below at the nitty gritty level.

Reply:
OK, I think I've added a clarifying section at the beginning of all this nitty-gritty stuff. :)